File: internal/models/capability.go
// agency-gateway/internal/models/capability.go
package models

import (
	"fmt"
	"unicode"

	agencysecurity "github.com/geoffbelknap/agency/internal/security"
)

// CapabilityKind is the type of a capability entry.
// Valid values: "mcp-server", "skill", "service".
type CapabilityKind = string

// CapabilityState is the operational state of a capability.
// Valid values: "available", "restricted", "disabled".
type CapabilityState = string

// ToolApproval is the approval policy for a tool.
// Valid values: "available", "ask-once", "ask-always", "denied".
type ToolApproval = string

// CapabilityRequirements lists prerequisites for a capability.
type CapabilityRequirements struct {
	RuntimePackages []string `yaml:"runtime_packages"`
	Network         []string `yaml:"network"`
	Capabilities    []string `yaml:"capabilities"`
}

// CapabilityPermissions declares the permission level required by a capability.
type CapabilityPermissions struct {
	// Filesystem is one of "none", "read-only", "read-write". Defaults to "none".
	Filesystem string `yaml:"filesystem" validate:"oneof=none read-only read-write" default:"none"`
	Network    bool   `yaml:"network" default:"false"`
	Execution  bool   `yaml:"execution" default:"false"`
}

// CapabilityIntegrity holds integrity verification data for a capability.
type CapabilityIntegrity struct {
	SHA256     *string `yaml:"sha256"`
	SignedBy   *string `yaml:"signed_by"`
	Signature  *string `yaml:"signature"`
	VerifiedAt *string `yaml:"verified_at"`
}

// MCPServerSpec defines how to launch an MCP server capability.
type MCPServerSpec struct {
	Command string            `yaml:"command" validate:"required"`
	Args    []string          `yaml:"args"`
	Env     map[string]string `yaml:"env"`
}

// CapabilityEntry is the definition of a capability in the capability registry.
type CapabilityEntry struct {
	Kind        CapabilityKind         `yaml:"kind" validate:"required,oneof=mcp-server skill service"`
	Name        string                 `yaml:"name" validate:"required"`
	Version     string                 `yaml:"version" default:"0.1.0"`
	DisplayName string                 `yaml:"display_name"`
	Description string                 `yaml:"description"`
	Source      string                 `yaml:"source" default:"local"`
	Publisher   string                 `yaml:"publisher"`
	Integrity   CapabilityIntegrity    `yaml:"integrity"`
	Requires    CapabilityRequirements `yaml:"requires"`
	Permissions CapabilityPermissions  `yaml:"permissions"`
	Spec        *MCPServerSpec         `yaml:"spec"`
	ServiceRef  *string                `yaml:"service_ref"`
	SkillPath   *string                `yaml:"skill_path"`
	Tags        []string               `yaml:"tags"`
}

// Validate checks that the capability name is alphanumeric with hyphens or underscores.
func (c *CapabilityEntry) Validate() error {
	for _, r := range c.Name {
		if !unicode.IsLetter(r) && !unicode.IsDigit(r) && r != '-' && r != '_' {
			return fmt.Errorf("capability name must be alphanumeric with hyphens or underscores")
		}
	}
	return nil
}

// ToolPolicy defines the approval policy for a specific tool.
type ToolPolicy struct {
	Approval ToolApproval `yaml:"approval" validate:"oneof=available ask-once ask-always denied" default:"available"`
}

// CapabilityAuth defines how to authenticate a capability.
type CapabilityAuth struct {
	Env      string            `yaml:"env" validate:"required"`
	InjectAs string            `yaml:"inject_as"`
	Agents   map[string]string `yaml:"agents"`
}

// CapabilityConfig is the runtime configuration for a capability in capabilities.yaml.
type CapabilityConfig struct {
	// State is one of "available", "restricted", "disabled". Defaults to "available".
	State  CapabilityState         `yaml:"state" validate:"oneof=available restricted disabled" default:"available"`
	Agents []string                `yaml:"agents"`
	Auth   *CapabilityAuth         `yaml:"auth"`
	Tools  map[string]ToolApproval `yaml:"tools"`
}

// CapabilitiesFile is the top-level schema for capabilities.yaml.
type CapabilitiesFile struct {
	Capabilities map[string]CapabilityConfig `yaml:"capabilities"`
}

// ToolApprovalRecord records an operator approval decision for a specific tool call.
type ToolApprovalRecord struct {
	Capability string                        `yaml:"capability" validate:"required"`
	Tool       string                        `yaml:"tool" validate:"required"`
	Agent      string                        `yaml:"agent" validate:"required"`
	Approved   bool                          `yaml:"approved"`
	Status     agencysecurity.ApprovalStatus `yaml:"status,omitempty"`
	ApprovedBy string                        `yaml:"approved_by" default:"operator"`
	ApprovedAt string                        `yaml:"approved_at"`
}

func (r ToolApprovalRecord) ApprovalStatus() agencysecurity.ApprovalStatus {
	if r.Status != "" {
		return r.Status
	}
	if r.Approved {
		return agencysecurity.ApprovalApproved
	}
	return agencysecurity.ApprovalDenied
}

// CapabilityPolicy declares which capabilities an agent is permitted to use.
type CapabilityPolicy struct {
	Required  []string `yaml:"required"`
	Available []string `yaml:"available"`
	Denied    []string `yaml:"denied"`
	Enabled   []string `yaml:"enabled"`
}

File: internal/images/contexts/_shared/models/capability.py
"""Capability registry models.

Defines schemas for capability entries (MCP servers, skills, services)
and the central capabilities.yaml configuration.
"""

from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator


CapabilityKind = Literal["mcp-server", "skill", "service"]
CapabilityState = Literal["available", "restricted", "disabled"]
ToolApproval = Literal["available", "ask-once", "ask-always", "denied"]
ApprovalStatus = Literal["pending", "routed", "approved", "rejected", "denied", "canceled"]


class CapabilityRequirements(BaseModel):
    """What a capability needs from the workspace."""

    model_config = ConfigDict(extra="forbid")

    runtime_packages: list[str] = []
    network: list[str] = []
    capabilities: list[str] = []


class CapabilityPermissions(BaseModel):
    """What a capability can do — used for policy validation."""

    model_config = ConfigDict(extra="forbid")

    filesystem: Literal["none", "read-only", "read-write"] = "none"
    network: bool = False
    execution: bool = False


class CapabilityIntegrity(BaseModel):
    """Integrity verification for marketplace packages."""

    model_config = ConfigDict(extra="forbid")

    sha256: str | None = None
    signed_by: str | None = None
    signature: str | None = None
    verified_at: str | None = None


class MCPServerSpec(BaseModel):
    """MCP server-specific configuration."""

    model_config = ConfigDict(extra="forbid")

    command: str
    args: list[str] = []
    env: dict[str, str] = {}


class CapabilityEntry(BaseModel):
    """A capability in the registry.

    Common envelope for all capability types: mcp-server, skill, service.
    """

    model_config = ConfigDict(extra="forbid")

    kind: CapabilityKind
    name: str
    version: str = "0.1.0"
    display_name: str = ""
    description: str = ""
    source: str = "local"
    publisher: str = ""

    integrity: CapabilityIntegrity = CapabilityIntegrity()
    requires: CapabilityRequirements = CapabilityRequirements()
    permissions: CapabilityPermissions = CapabilityPermissions()

    # Type-specific spec — present for mcp-server entries
    spec: MCPServerSpec | None = None

    # For service entries — reference to a service definition file
    service_ref: str | None = None

    # For skill entries — path to the skill directory
    skill_path: str | None = None

    tags: list[str] = []

    @field_validator("name")
    @classmethod
    def name_is_valid(cls, v: str) -> str:
        if not v.replace("-", "").replace("_", "").isalnum():
            raise ValueError(
                "Capability name must be alphanumeric with hyphens or underscores"
            )
        return v


# ---------------------------------------------------------------------------
# Central capabilities.yaml configuration
# ---------------------------------------------------------------------------


class ToolPolicy(BaseModel):
    """Per-tool approval policy within a capability."""

    model_config = ConfigDict(extra="forbid")

    approval: ToolApproval = "available"


class CapabilityAuth(BaseModel):
    """Authentication configuration for a capability.

    Controls which API key/token is injected into the MCP server
    environment at start time.

    - env: env var name in the secrets file holding the default key
    - inject_as: env var name inside the MCP server process
    - agents: per-agent env var overrides (agent_name -> env_var_name)
    """

    model_config = ConfigDict(extra="forbid")

    env: str                           # e.g. BRAVE_API_KEY
    inject_as: str = ""                # defaults to same as env if empty
    agents: dict[str, str] = {}        # e.g. {"atlas": "BRAVE_API_KEY_ATLAS"}


class CapabilityConfig(BaseModel):
    """Configuration for a single capability in capabilities.yaml.

    Controls who can use it, what auth is injected, and what
    approval is needed per-tool.
    """

    model_config = ConfigDict(extra="forbid")

    state: CapabilityState = "available"
    agents: list[str] = []  # only used when state=restricted
    auth: CapabilityAuth | None = None
    tools: dict[str, ToolApproval] = {}  # per-tool approval overrides

    @field_validator("agents")
    @classmethod
    def agents_only_with_restricted(cls, v, info):
        # Allow agents list regardless — validation at usage time
        return v


class CapabilitiesFile(BaseModel):
    """Schema for ~/.agency/capabilities.yaml.

    Central configuration controlling which capabilities are active,
    who can use them, and what tool-level approval is required.
    """

    model_config = ConfigDict(extra="forbid")

    capabilities: dict[str, CapabilityConfig] = {}


# ---------------------------------------------------------------------------
# Tool approval state tracking
# ---------------------------------------------------------------------------


class ToolApprovalRecord(BaseModel):
    """Tracks operator approval for ask-once tools."""

    model_config = ConfigDict(extra="forbid")

    capability: str
    tool: str
    agent: str
    approved: bool
    status: ApprovalStatus | None = None
    approved_by: str = "operator"
    approved_at: str = ""


# ---------------------------------------------------------------------------
# Legacy compatibility (kept for policy.yaml migration)
# ---------------------------------------------------------------------------


class CapabilityPolicy(BaseModel):
    """Legacy capability policy section in policy.yaml.

    Retained for backward compatibility. New deployments should use
    capabilities.yaml instead.
    """

    model_config = ConfigDict(extra="forbid")

    required: list[str] = []
    available: list[str] = []
    denied: list[str] = []
    enabled: list[str] = []
